colossalai/utils/checkpointing.py

```diff
@@ -1,212 +1,253 @@
-import os
-import os.path as osp
-import re
-from typing import Tuple
-from pathlib import Path
+from collections import OrderedDict
+from itertools import chain
 
 import torch
-
-from colossalai.context import Config
+import torch.distributed as dist
+from colossalai.communication.collective import scatter_object_list
 from colossalai.context.parallel_mode import ParallelMode
 from colossalai.core import global_context as gpc
+from torch.nn.modules.module import _EXTRA_STATE_KEY_SUFFIX
 
-__all__ = [
-    'get_checkpoint_path', 'get_latest_checkpoint_path', 'get_latest_checkpoint_pattern', 'save_checkpoint',
-    'load_checkpoint'
-]
-
-
-def unwrap_config(config: Config):
-    """Unwrap Config objects to normal dicts
-    """
-    config_dict = dict()
-    for k, v in config.items():
-        if isinstance(v, dict):
-            config_dict[k] = unwrap_config(v)
-        else:
-            config_dict[k] = v
-
-    return config_dict
+from .common import is_using_pp
 
+__all__ = ["save_checkpoint", "load_checkpoint"]
 
-def _get_ranks_name():
-    # tensor parallel
-    tp_local_rank = 0
-    if gpc.is_initialized(ParallelMode.TENSOR):
-        tp_local_rank = gpc.get_local_rank(ParallelMode.TENSOR)
 
-    # pipeline parallel
-    pp_local_rank = 0
-    if gpc.is_initialized(ParallelMode.PIPELINE):
-        pp_local_rank = gpc.get_local_rank(ParallelMode.PIPELINE)
+def broadcast_state_dict(state_dict, parallel_mode):
+    state_dict = [state_dict.copy() if isinstance(state_dict, dict) else state_dict]
```

Comment:
what type could state_dict be?
isinstance(state_dict, dict) may not be enough.
If it is a List or Set you also need to assign a copy?

Reply:
@feifeibear Only the source rank has the dict (an `OrderedDict` loaded from file) to broadcast. Otherwise, we don't care what type it is, since it is going to be replaced by broadcast anyway.